### PR TITLE
feat(kuma-cp): add support for Gateway in MeshLoadBalancingStrategy

### PR DESCRIPTION
**Pull request by Automaat**
### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/policies/core/xds/endpoints.go [hunk 1]
```diff
@@ -38,6 +39,30 @@ func GatherEndpoints(rs *xds.ResourceSet) EndpointMap {
 	return em
 }
 
+func GatherGatewayEndpoints(rs *xds.ResourceSet) EndpointMap {
+	em := EndpointMap{}
+	for _, res := range rs.Resources(envoy_resource.EndpointType) {
```
Comment (lukidzi) [inline pkg/plugins/policies/core/xds/endpoints.go:44]: maybe we can make method `gatherEndpoints` which has parameter `gatherEndpoints(rs *xds.ResourceSet, origin string)`

File: pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/plugin_test.go [hunk 2]
```diff
@@ -1345,13 +1356,20 @@ var _ = Describe("MeshLoadBalancingStrategy", func() {
...
 			Expect(getResourceYaml(generatedResources.ListOf(envoy_resource.ListenerType))).
 				To(matchers.MatchGoldenYAML(filepath.Join("testdata", fmt.Sprintf("%s.gateway_listener.golden.yaml", given.name))))
 			Expect(getResourceYaml(generatedResources.ListOf(envoy_resource.RouteType))).
 				To(matchers.MatchGoldenYAML(filepath.Join("testdata", fmt.Sprintf("%s.gateway_route.golden.yaml", given.name))))
 		},
 		Entry("basic outbound cluster", gatewayTestCase{
 			name: "basic",
+			endpointMap: map[core_xds.ServiceName][]core_xds.Endpoint{
+				"backend": {
+					createEndpointWith("test-zone", "192.168.1.1", map[string]string{}),
```
Comment (lukidzi) [inline pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/plugin_test.go:1370]: Can we add an endpoint in the local zone and in a different zone to check if it works correctly for `disabled`?

**Review by slonka — COMMENTED** [review]
If we could squeeze in an e2e tests with this that would be great.

File: pkg/plugins/policies/core/xds/endpoints.go [hunk 3]
```diff
@@ -6,58 +6,46 @@ import (
...
 
-		cluster := res.Resource.(*clusterv3.Cluster)
-		serviceName := tags.ServiceFromClusterName(cluster.Name)
-		if cluster.LoadAssignment != nil {
-			em[serviceName] = append(em[serviceName], cluster.LoadAssignment)
-		}
-	}
-	return em
+func GatherGatewayEndpoints(rs *xds.ResourceSet) EndpointMap {
+	return gatherEndpoints(rs, metadata.OriginGateway)
```
[the remark this answers is not in the document]
Comment (Automaat) [reply pkg/plugins/policies/core/xds/endpoints.go:22]: yes

File: pkg/test/xds/meshcontext.go [hunk 4]
```diff
@@ -33,3 +33,18 @@ func CreateSampleMeshContextWith(resources xds_context.Resources) xds_context.Co
 		ControlPlane: &xds_context.ControlPlaneContext{CLACache: &DummyCLACache{}, Zone: "test-zone"},
 	}
 }
+
+func CreateSampleMeshContextWithEndpoints(resources xds_context.Resources, endpointMap core_xds.EndpointMap) xds_context.Context {
```
[the remark this answers is not in the document]
Comment (Automaat) [reply pkg/test/xds/meshcontext.go:37]: I agree, but I don't want to change too much here just before release

File: pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/plugin.go [hunk 5]
```diff
@@ -204,7 +206,7 @@ func (p plugin) configureGateway(
 				}
 
 				serviceName := dest.Destination[mesh_proto.ServiceTag]
-				if err := configureEndpoints(proxy.Dataplane.Spec.TagSet(), cluster, endpoints[serviceName], serviceName, *conf, rs, proxy.Zone, proxy.APIVersion, egressEnabled, generator.OriginOutbound); err != nil {
+				if err := configureEndpoints(proxy.Dataplane.Spec.TagSet(), cluster, endpoints[serviceName], serviceName, *conf, rs, proxy.Zone, proxy.APIVersion, egressEnabled, metadata.OriginGateway); err != nil {
```
[the remark this answers is not in the document]
Comment (Automaat) [reply pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/plugin.go]: yes

File: pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/plugin_test.go (hunk 2 above)
Comment (Automaat) [reply pkg/plugins/policies/meshloadbalancingstrategy/plugin/v1alpha1/plugin_test.go:1370]: added test

File: test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/kumahq/kuma/test/framework/deployments/democlient"
+	"github.com/kumahq/kuma/test/framework/deployments/testserver"
+	"github.com/kumahq/kuma/test/framework/envs/multizone"
+)
+
+func LocalityAwareLBGateway() {
+	const mesh = "locality-aware-lb-gateway"
+	const namespace = "locality-aware-lb-gateway"
+
+	meshLoadBalancingStrategyDemoClient := fmt.Sprintf(`
```
Comment (lukidzi) [inline test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go:22]: nit: name points that affect demo-client but affect all dps that communicate with test-server

File: pkg/test/xds/cla_cache.go [hunk 7]
```diff
@@ -15,7 +15,7 @@ type DummyCLACache struct {
 }
 
 func (d *DummyCLACache) GetCLA(ctx context.Context, meshName, meshHash string, cluster envoy_common.Cluster, apiVersion core_xds.APIVersion, endpointMap core_xds.EndpointMap) (proto.Message, error) {
-	return endpoints.CreateClusterLoadAssignment(cluster.Service(), d.OutboundTargets[cluster.Service()]), nil
+	return endpoints.CreateClusterLoadAssignment(cluster.Name(), d.OutboundTargets[cluster.Service()]), nil
```
Comment (slonka) [inline pkg/test/xds/cla_cache.go:18]: how is this related? was it wrong always?

File: test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+				HaveKeyWithValue(Equal(`test-server-zone-1`), BeNumerically("~", 25, 10)),
+			),
+		)
+
+		// apply lb policy
+		// kuma-4 - priority 0, kuma-5 - priority 1, kuma-1 - priority 2
+		Expect(multizone.Global.Install(YamlUniversal(meshLoadBalancingStrategyDemoClient))).To(Succeed())
+
+		Eventually(func() (map[string]int, error) {
+			gatewayIP := multizone.UniZone1.GetApp("edge-gateway").GetIP()
```
Comment (slonka) [inline test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go:159]: nitpick: I guess the IP won't change right?

**Review by slonka — APPROVED** [review]
looks good, but I don't understand dla CLA cache change

File: pkg/test/xds/cla_cache.go (hunk 7 above)
Comment (Automaat) [reply pkg/test/xds/cla_cache.go:18]: yes it was incorrect allways

Comment (lukidzi) [reply pkg/test/xds/cla_cache.go:18]: Also, it's `test/cla_cache.go` resource - not cla_cache change

**Review by lukidzi — APPROVED** [review]

File: pkg/test/xds/cla_cache.go (hunk 7 above)
Comment (slonka) [reply pkg/test/xds/cla_cache.go:18]: yeah I know, I'm just worried about this breaking something else

File: test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go (hunk 8 above)
Comment (Automaat) [reply test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go:159]: It should not, changed

File: pkg/test/xds/cla_cache.go (hunk 7 above)
Comment (Automaat) [reply pkg/test/xds/cla_cache.go:18]: If it breaks something we will see this in tests, but method `CreateClusterLoadAssignment` should get cluster name and in old version it got service name instead of cluster name

File: test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go (hunk 6 above)
Comment (Automaat) [reply test/e2e_env/multizone/localityawarelb/locality_aware_multizone_hybrid_gateway.go]: changed
